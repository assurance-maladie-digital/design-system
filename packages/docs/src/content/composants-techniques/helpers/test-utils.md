---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
