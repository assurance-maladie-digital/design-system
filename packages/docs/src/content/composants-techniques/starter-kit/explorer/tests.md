Le dossier `tests` contient les tests d’interface et les tests unitaires globaux de l’application.

<doc-alert type="warning">
Il n’est pas recommandé de renommer ce dossier.
</doc-alert>
