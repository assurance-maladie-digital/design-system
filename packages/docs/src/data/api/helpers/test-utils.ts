import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
