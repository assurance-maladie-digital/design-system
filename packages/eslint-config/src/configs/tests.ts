import { type FlatESLintConfigItem } from 'eslint-define-config';
import { pluginVitest } from '../plugins';
import { GLOB_TESTS } from '../globs';
import { isInEditor } from '../env';

export const tests: FlatESLintConfigItem[] = [
	{
		plugins: {
			tests: pluginVitest,
		},
	},
	{
		files: GLOB_TESTS,
		rules: {
			'tests/consistent-test-it': [
				'error',
				{
					fn: 'it',
					withinDescribe: 'it',
				},
			],
			'tests/expect-expect': 'error',
			'tests/no-alias-methods': 'error',
			'tests/no-commented-out-tests': isInEditor ? 'off' : 'error',
			'tests/no-conditional-expect': 'error',
			'tests/no-disabled-tests': 'error',
			'tests/no-done-callback': 'error',
			'tests/no-duplicate-hooks': 'error',
			'tests/no-focused-tests': isInEditor ? 'off' : 'error',
			'tests/no-identical-title': 'error',
			'tests/no-interpolation-in-snapshots': 'error',
			'tests/no-standalone-expect': 'error',
			'tests/no-test-return-statement': 'error',
			'tests/prefer-comparison-matcher': 'error',
			'tests/prefer-equality-matcher': 'error',
			'tests/prefer-hooks-in-order': 'error',
			'tests/prefer-hooks-on-top': 'error',
			'tests/prefer-mock-promise-shorthand': 'error',
			'tests/prefer-spy-on': 'error',
			'tests/prefer-to-be-falsy': 'error',
			'tests/prefer-to-be-object': 'error',
			'tests/prefer-to-be-truthy': 'error',
			'tests/prefer-to-be': 'error',
			'tests/prefer-to-contain': 'error',
			'tests/prefer-to-have-length': 'error',
			'tests/prefer-todo': 'error',
			'tests/require-hook': 'error',
			'tests/require-top-level-describe': 'error',
			'tests/valid-describe-callback': 'error',
			'tests/valid-expect': 'error',
			'tests/valid-title': 'error',
		},
	},
];
