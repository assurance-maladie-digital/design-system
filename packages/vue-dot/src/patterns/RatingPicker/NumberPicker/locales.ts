export const locales = {
	ariaLabel: (index: number, length: number): string => `Note de ${index} sur ${length}.`
};
