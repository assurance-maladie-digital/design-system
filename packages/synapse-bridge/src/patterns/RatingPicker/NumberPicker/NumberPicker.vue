<script lang="ts">
import { defineComponent } from "vue";
import type { PropType } from "vue";
import { RatingMixin } from "../RatingMixin";
import { locales } from "./locales";
interface SelectItem {
	title: string;
	value: number;
}
export default defineComponent({
	mixins: [RatingMixin],
	emits: ["update:modelValue"],
	props: {
		length: {
			type: Number,
			default: 10,
		},
		itemLabels: {
			type: Array as PropType<string[]>,
			default: () => [],
		},
	},
	data() {
		return {
			locales,
		};
	},
	computed: {
		selectItems(): SelectItem[] {
			return [...Array(this.length)].map((_, index) => ({
				title: `${index + 1}`,
				value: index + 1
			}));
		},
		shouldDisplayLabels(): boolean {
			return this.itemLabels.length === 2;
		},
		hasAnswered(): boolean {
			return this.modelValue !== -1;
		}
	},
	methods: {
		setValue(value: string | number): void {
			this.emitInputEvent(value);
		}
	}
});
</script>

<template>
	<fieldset class="vd-number-picker">
		<VSelect
			v-if="$vuetify.display.xs"
			:modelValue="modelValue === -1 ? null : modelValue"
			:label="label ?? ''"
			:disabled="readonly || hasAnswered"
			:items="selectItems"
			hide-details
			variant="outlined"
			class="vd-form-input"
			@update:modelValue="setValue"
		/>
		<template v-else>
			<legend class="text-h6 mb-6">
				<slot name="label">
					{{ label }}
				</slot>
			</legend>
			<div v-if="!hasAnswered" class="d-inline-block">
				<VRating
					:modelValue="modelValue"
					:length="length"
					:readonly="readonly || hasAnswered"
					class="d-flex flex-wrap max-width-none"
					@update:modelValue="setValue"
				>
					<template #item="{ index }">
						<VBtn
							:aria-label="locales.ariaLabel(index + 1, length)"
							:disabled="readonly"
							size="x-small"
							variant="outlined"
							color="primary"
							height="36px"
							class="text-body-2 pa-0 mr-2"
						>
							{{ index + 1 }}
						</VBtn>
					</template>
				</VRating>
				<div
					v-if="shouldDisplayLabels"
					class="d-flex justify-space-between mt-1"
				>
					<span
						:aria-label="`${locales.ariaLabel(1, length)} ${
							itemLabels[0]
						}.`"
						class="text-caption"
						v-text="itemLabels[0]"
					/>
					<span
						:aria-label="`${locales.ariaLabel(length, length)} ${
							itemLabels[1]
						}.`"
						class="text-caption mr-2"
						v-text="itemLabels[1]"
					/>
				</div>
			</div>
			<p
				v-else
				:aria-label="locales.ariaLabel(modelValue, length)"
				class="mb-0 d-flex align-center"
			>
				<VBtn
					aria-hidden="true"
					:disabled="true"
					size="x-small"
					variant="outlined"
					color="primary"
					height="36px"
					class="vd-btn-answer text-body-2 mr-1 pa-0"
				>
					{{ modelValue }}
				</VBtn>
				/ {{ length }}
			</p>
		</template>
	</fieldset>
</template>
<style lang="scss" scoped>
@import "@cnamts/design-tokens/dist/tokens";
.vd-number-picker {
	border: 0;
}
.vd-btn-answer.v-btn.v-btn--disabled {
	opacity: 1;
}
</style>
