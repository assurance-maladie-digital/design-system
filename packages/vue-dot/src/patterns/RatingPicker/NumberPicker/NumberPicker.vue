<template>
	<fieldset class="vd-number-picker">
		<VSelect
			v-if="$vuetify.breakpoint.xs"
			:value="value"
			:label="label"
			:disabled="readonly || hasAnswered"
			:items="selectItems"
			hide-details
			outlined
			class="vd-form-input"
			@change="setValue"
		/>

		<template v-else>
			<legend class="text-h6 mb-6">
				{{ label }}
			</legend>

			<div
				v-if="!hasAnswered"
				class="d-inline-block"
			>
				<VRating
					:value="value"
					:length="length"
					:readonly="readonly || hasAnswered"
					class="d-flex flex-wrap mx-n1 max-width-none"
					@input="setValue"
				>
					<template #item="{ index, click }">
						<VBtn
							:aria-label="locales.ariaLabel(index + 1, length)"
							:disabled="readonly"
							x-small
							outlined
							color="primary"
							height="36px"
							class="text-body-2 mx-1 pa-0"
							@click="click"
						>
							{{ index + 1 }}
						</VBtn>
					</template>
				</VRating>

				<div
					v-if="shouldDisplayLabels"
					class="d-flex justify-space-between mt-1"
				>
					<span
						:aria-label="`${locales.ariaLabel(1, length)} ${itemLabels[0]}.`"
						class="text-caption"
						v-text="itemLabels[0]"
					/>

					<span
						:aria-label="`${locales.ariaLabel(length, length)} ${itemLabels[1]}.`"
						class="text-caption"
						v-text="itemLabels[1]"
					/>
				</div>
			</div>

			<p
				v-else
				:aria-label="locales.ariaLabel(value, length)"
				class="mb-0"
			>
				<VBtn
					aria-hidden="true"
					disabled
					x-small
					outlined
					color="primary"
					height="36px"
					class="vd-btn-answer text-body-2 mr-1 pa-0"
				>
					{{ value }}
				</VBtn>

				/ {{ length }}
			</p>
		</template>
	</fieldset>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { RatingMixin } from '../RatingMixin';

	import { locales } from './locales';

	interface SelectItem {
		text: string;
		value: number;
	}

	const Props = Vue.extend({
		props: {
			length: {
				type: Number,
				default: 10
			},
			itemLabels: {
				type: Array as PropType<string[]>,
				default: () => []
			}
		}
	});

	const MixinsDeclaration = mixins(Props, RatingMixin);

	@Component
	export default class NumberPicker extends MixinsDeclaration {
		locales = locales;

		get selectItems(): SelectItem[] {
			return [...Array(this.length)].map((_, index) => ({
				text: `${index + 1}`,
				value: index + 1
			}));
		}

		get shouldDisplayLabels(): boolean {
			return this.itemLabels.length === 2;
		}

		get hasAnswered(): boolean {
			return this.value !== -1;
		}

		setValue(value: number): void {
			this.emitInputEvent(value);
		}
	}
</script>

<style lang="scss" scoped>
	@import '@cnamts/design-tokens/dist/tokens';

	.vd-btn-answer.v-btn.v-btn--disabled {
		color: $vd-primary !important;
	}
</style>
