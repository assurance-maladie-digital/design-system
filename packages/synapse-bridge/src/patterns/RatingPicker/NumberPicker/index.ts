import NumberPicker from './NumberPicker.vue';

export default NumberPicker;
