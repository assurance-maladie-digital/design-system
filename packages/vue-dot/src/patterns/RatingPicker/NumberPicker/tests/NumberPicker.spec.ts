import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';

import NumberPicker from '../';

let wrapper: Wrapper<Vue>;

describe('NumberPicker', () => {
	it('renders correctly', () => {
		wrapper = mountComponent(NumberPicker,{
			propsData: {
				label: 'Pourriez-vous donner une note ?'
			}
		}, true);

		expect(wrapper).toMatchSnapshot();
	});
});
