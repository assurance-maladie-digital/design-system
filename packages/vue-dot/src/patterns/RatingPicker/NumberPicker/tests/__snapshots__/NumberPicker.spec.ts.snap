// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberPicker renders correctly 1`] = `
<fieldset class="vd-number-picker">
  <legend class="text-h6 mb-6">
    Pourriez-vous donner une note ?
  </legend>
  <div class="d-inline-block">
    <div class="v-rating d-flex flex-wrap mx-n1 max-width-none"><button type="button" class="text-body-2 mx-1 pa-0 v-btn v-btn--outlined theme--light v-size--x-small primary--text" style="height: 36px;" aria-label="Note de 1 sur 10."><span class="v-btn__content">
						1
					</span></button><button type="button" class="text-body-2 mx-1 pa-0 v-btn v-btn--outlined theme--light v-size--x-small primary--text" style="height: 36px;" aria-label="Note de 2 sur 10."><span class="v-btn__content">
						2
					</span></button><button type="button" class="text-body-2 mx-1 pa-0 v-btn v-btn--outlined theme--light v-size--x-small primary--text" style="height: 36px;" aria-label="Note de 3 sur 10."><span class="v-btn__content">
						3
					</span></button><button type="button" class="text-body-2 mx-1 pa-0 v-btn v-btn--outlined theme--light v-size--x-small primary--text" style="height: 36px;" aria-label="Note de 4 sur 10."><span class="v-btn__content">
						4
					</span></button><button type="button" class="text-body-2 mx-1 pa-0 v-btn v-btn--outlined theme--light v-size--x-small primary--text" style="height: 36px;" aria-label="Note de 5 sur 10."><span class="v-btn__content">
						5
					</span></button><button type="button" class="text-body-2 mx-1 pa-0 v-btn v-btn--outlined theme--light v-size--x-small primary--text" style="height: 36px;" aria-label="Note de 6 sur 10."><span class="v-btn__content">
						6
					</span></button><button type="button" class="text-body-2 mx-1 pa-0 v-btn v-btn--outlined theme--light v-size--x-small primary--text" style="height: 36px;" aria-label="Note de 7 sur 10."><span class="v-btn__content">
						7
					</span></button><button type="button" class="text-body-2 mx-1 pa-0 v-btn v-btn--outlined theme--light v-size--x-small primary--text" style="height: 36px;" aria-label="Note de 8 sur 10."><span class="v-btn__content">
						8
					</span></button><button type="button" class="text-body-2 mx-1 pa-0 v-btn v-btn--outlined theme--light v-size--x-small primary--text" style="height: 36px;" aria-label="Note de 9 sur 10."><span class="v-btn__content">
						9
					</span></button><button type="button" class="text-body-2 mx-1 pa-0 v-btn v-btn--outlined theme--light v-size--x-small primary--text" style="height: 36px;" aria-label="Note de 10 sur 10."><span class="v-btn__content">
						10
					</span></button></div>
    <!---->
  </div>
</fieldset>
`;
